### PR TITLE
Complete the request a referee flow

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

Previous PRs are:
#3013 - Adds the new start and type page
#3016 - Migration to allow the references email address column to be NULL
#3022 - Breaks the new flow into a separate controller
#3035 - Up to adding a referees name
#3040 - Add referees email
#3044 - Adds the relationship page 
#3094 - Adds the unsubmitted review page

This is the final PR for the new decoupled references flow that links into the Candidate Name page steve added.

## Changes proposed in this pull request

1. It links the flow in with Steve's work
2. It moves validations from the Review controller to the SubmitRefereeForm and adds tests for them
3. Renames the RefereeSubmitForm to SubmitRefereeForm

## Link to Trello card

https://trello.com/c/npGvdYqK/2213-%F0%9F%92%94-dev-request-a-referee-flow-mvp

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/controllers/candidate_interface/decoupled_references/review_controller.rb [hunk 1]
```diff
@@ -11,29 +11,15 @@ def show
...
-            # TODO: Link up with Steve's work
-          else
-            # TODO: Once the forms are moved into objects this won't be necessary
-            # It can just call valid? and render the errors.
-            flash[:warning] = 'You must complete all of your referees details before your reference request can be submitted'
-
-            render :unsubmitted
-          end
+        if @submit_reference_form.send_request?
+          redirect_to candidate_interface_decoupled_references_create_candidate_name_path(@reference.id)
```
Comment (stevehook) [inline app/controllers/candidate_interface/decoupled_references/review_controller.rb:22]: Does this work as the `create` action is a post?

Comment (davidgisbey) [reply app/controllers/candidate_interface/decoupled_references/review_controller.rb:22]: Good point. As discussed i've added some logic to the new action. The duplication is taken care of by Malcolm's PR 
image

Comment (malcolmbaig) [reply app/controllers/candidate_interface/decoupled_references/review_controller.rb:22]: I reckon we should do those checks on CandidateNameForm as part of ReviewController#submit (can be refactored into a private method). 

Something like:
```
if @reference_candidate_name_form.valid?
  CandidateInterface::DecoupledReferences::RequestReference.call(@reference, flash)
else
  redirect_to candidate_interface_decoupled_references_new_candidate_name_path
end
```

That way we aren't triggering side effects via a GET request.

**Review by malcolmbaig — CHANGES_REQUESTED** [review]
LGTM, just one comment about the interaction between the Review and CandidateName controller. Using the `DecoupledReferences::RequestReference` service object that's now in master will also ensure we don't send repeat emails via a stale form.

File: app/controllers/candidate_interface/decoupled_references/review_controller.rb (hunk 1 above)
Comment (davidgisbey) [reply app/controllers/candidate_interface/decoupled_references/review_controller.rb]: Have done 👍

**Review by malcolmbaig — APPROVED** [review]
A system spec needs updating, and there are a couple of typos in commits 560bafd0a5d989ac18bf13a7170edec7d521e6f6 and b9f8e313fb6131a6cd5989a70dc67689b8d306b5, but otherwise this is good to go.

**Review by stevehook — COMMENTED** [review]
Testing manually I still get prompted for my name if I select 'No, not at the moment'

**Review by tvararu — APPROVED** [review]
Haven't tested but code LGTM!

**Review by malcolmbaig — CHANGES_REQUESTED** [review]
Good spot Steve! https://github.com/DFE-Digital/apply-for-teacher-training/pull/3110#pullrequestreview-506602141

**Comment by davidgisbey** [comment]
Yea just fixed it.

File: app/controllers/candidate_interface/decoupled_references/review_controller.rb [hunk 2]
```diff
@@ -11,31 +11,21 @@ def show
...
-        else
+        @candidate_name_form = Reference::CandidateNameForm.build_from_reference(@reference)
+
+        if @submit_reference_form.send_request? && @candidate_name_form.valid?
+          CandidateInterface::DecoupledReferences::RequestReference.call(@reference, flash)
+
           redirect_to candidate_interface_decoupled_references_review_path
+        else
+          redirect_to candidate_interface_decoupled_references_new_candidate_name_path(@reference.id)
         end
```
Comment (stevehook) [inline app/controllers/candidate_interface/decoupled_references/review_controller.rb]:
```
       if !@candidate_name_form.valid?
           redirect_to candidate_interface_decoupled_references_review_path
         else
           CandidateInterface::DecoupledReferences::RequestReference.call(@reference, flash) if @submit_reference_form.send_request?

           redirect_to candidate_interface_decoupled_references_new_candidate_name_path(@reference.id)
         end
```

File: app/controllers/candidate_interface/decoupled_references/review_controller.rb [hunk 3]
```diff
@@ -11,31 +11,21 @@ def show
...
       end
```
Comment (stevehook) [inline app/controllers/candidate_interface/decoupled_references/review_controller.rb]: Might be a neater way to express it but I think this is the logic needed:
```
if @submit_reference_form.send_request? && !@candidate_name_form.valid?
  redirect_to candidate_interface_decoupled_references_new_candidate_name_path(@reference.id)
elsif @submit_reference_form.send_request?
  CandidateInterface::DecoupledReferences::RequestReference.call(@reference, flash)
  redirect_to candidate_interface_decoupled_references_review_path
else
  redirect_to candidate_interface_decoupled_references_review_path
end

```

Comment (davidgisbey) [reply app/controllers/candidate_interface/decoupled_references/review_controller.rb]: Yea that's neater than mine was 👍 thanks

**Review by malcolmbaig — APPROVED** [review]

**Review by stevehook — APPROVED** [review]

**Pull request merged**